security/landlock/supervisor.c

```diff
@@ -137,3 +164,270 @@ int landlock_supervisor_check(struct landlock_ruleset *ruleset,
 
 	return allow ? 0 : -EACCES;
 }
+
+/**
+ * check_cache_match - Check if cache entry matches current task
+ * @cache: Cache entry to check
+ *
+ * Returns true if cache entry matches current context.
+ */
+static bool check_cache_match(const struct landlock_supervisor_cache *cache)
+{
+	struct dentry *exe_dentry;
+	struct file *exe_file;
+	bool match = false;
+
+	switch (cache->match_type) {
+	case LANDLOCK_MATCH_ALL:
+		return true;
+
+	case LANDLOCK_MATCH_PID:
+		return cache->task == current;
+
+	case LANDLOCK_MATCH_PID_SUBTREE:
+		/* Check if current is cache->task or a descendant */
+		if (cache->task == current)
+			return true;
+		/* Simple check: if current's parent chain includes cache->task */
+		{
+			struct task_struct *p = current;
+			rcu_read_lock();
+			while (p && p != &init_task) {
+				if (p == cache->task) {
+					match = true;
+					break;
+				}
+				p = rcu_dereference(p->real_parent);
+			}
+			rcu_read_unlock();
+		}
+		return match;
+
+	case LANDLOCK_MATCH_EXEC:
+	case LANDLOCK_MATCH_EXEC_SUBTREE:
+		/* Get current executable */
+		exe_file = get_task_exe_file(current);
+		if (!exe_file)
+			return false;
+		exe_dentry = exe_file->f_path.dentry;
+		match = (exe_dentry == cache->exec_dentry);
+		fput(exe_file);
+		return match;
+
+	default:
+		return false;
+	}
+}
+
+/**
+ * landlock_supervisor_check_cache - Check if access is cached
+ * @supervisor: Supervisor to check cache
+ * @access: Requested access
+ *
+ * Returns 0 if cached and allowed, -EACCES if cached and denied,
+ * -ENOENT if not cached.
+ */
+static int landlock_supervisor_check_cache(struct landlock_supervisor *supervisor,
+					   u64 access)
+{
+	struct landlock_supervisor_cache *cache;
+	struct landlock_ruleset *merged = NULL;
+	unsigned long flags;
+	int ret = -ENOENT;
+	bool found_match = false;
+
+	if (!supervisor)
+		return -ENOENT;
+
+	spin_lock_irqsave(&supervisor->lock, flags);
+
+	/* Collect all matching cache entries and merge rulesets */
+	list_for_each_entry(cache, &supervisor->cache_list, list) {
+		if (!check_cache_match(cache))
+			continue;
+
+		found_match = true;
+
+		/* Merge cache ruleset into accumulated ruleset */
+		if (!merged) {
+			/* First match: use as base */
+			merged = cache->ruleset;
+			landlock_get_ruleset(merged);
+		} else {
+			/* Subsequent matches: merge */
+			struct landlock_ruleset *tmp;
+			tmp = landlock_merge_ruleset(merged, cache->ruleset);
```

Comment:
Does this merge return a single layer ruleset? Or does it merge them by stacking them into two layers?